// @boxyhq/saml20/dist/index.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const validateSignature_1 = require("./validateSignature");
const request_1 = require("./request");
const cert_1 = require("./cert");
const response_1 = require("./response");
const metadata_1 = require("./metadata");
const post_1 = require("./post");
const sign_1 = require("./sign");
const decrypt_1 = require("./decrypt");
const logout_1 = require("./logout");
exports.default = {
    parseMetadata: metadata_1.parseMetadata,
    createIdPMetadataXML: metadata_1.createIdPMetadataXML,
    createSPMetadataXML: metadata_1.createSPMetadataXML,
    createSAMLResponse: response_1.createSAMLResponse,
    request: request_1.request,
    parseSAMLRequest: request_1.parseSAMLRequest,
    decodeBase64: request_1.decodeBase64,
    parse: response_1.parse,
    validate: response_1.validate,
    PubKeyInfo: cert_1.PubKeyInfo,
    certToPEM: validateSignature_1.certToPEM,
    stripCertHeaderAndFooter: cert_1.stripCertHeaderAndFooter,
    createPostForm: post_1.createPostForm,
    sign: sign_1.sign,
    hasValidSignature: validateSignature_1.hasValidSignature,
    validateSignature: validateSignature_1.validateSignature,
    decryptXml: decrypt_1.decryptXml,
    parseIssuer: response_1.parseIssuer,
    WrapError: response_1.WrapError,
    parseLogoutResponse: logout_1.parseLogoutResponse,
    createLogoutRequest: logout_1.createLogoutRequest,
};
